File: Skyscraper-iOS-20251106/Skyscraper/Models/ChatModels.swift
//
//  ChatModels.swift
//  Skyscraper
//
//  Models for Bluesky Chat (Direct Messages)
//

import Foundation

// MARK: - Conversation

struct ConvoView: Codable, Identifiable, Equatable {
    let id: String
    let rev: String
    let members: [ConvoMember]
    let lastMessage: MessageUnion?
    let muted: Bool
    let unreadCount: Int

    static func == (lhs: ConvoView, rhs: ConvoView) -> Bool {
        lhs.id == rhs.id &&
        lhs.rev == rhs.rev &&
        lhs.unreadCount == rhs.unreadCount
    }

    var otherMembers: [ConvoMember] {
        // Filter out the current user
        members.filter { !$0.did.contains(ATProtoClient.shared.session?.did ?? "") }
    }

    var displayName: String {
        let others = otherMembers
        if others.count == 1, let member = others.first {
            return member.displayName ?? member.handle
        } else if others.count > 1 {
            return others.map { $0.displayName ?? $0.handle }.prefix(2).joined(separator: ", ")
        }
        return "Conversation"
    }

    var displayHandle: String? {
        let others = otherMembers
        if others.count == 1, let member = others.first {
            return "@\(member.handle)"
        }
        return nil
    }

    var avatarURL: URL? {
        otherMembers.first?.avatar.flatMap { URL(string: $0) }
    }
}

struct ConvoMember: Codable, Identifiable {
    let did: String
    let handle: String
    let displayName: String?
    let avatar: String?
    let associated: Associated?

    var id: String { did }

    struct Associated: Codable {
        let chat: ChatAssociated?

        struct ChatAssociated: Codable {
            let allowIncoming: String
        }
    }
}

// MARK: - Messages

struct MessageSender: Codable, Identifiable {
    let did: String

    var id: String { did }
}

enum MessageUnion: Codable {
    case messageView(MessageView)
    case deletedMessageView(DeletedMessageView)

    private enum CodingKeys: String, CodingKey {
        case type = "$type"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .type)

        let singleValueContainer = try decoder.singleValueContainer()

        switch type {
        case "chat.bsky.convo.defs#messageView":
            let messageView = try singleValueContainer.decode(MessageView.self)
            self = .messageView(messageView)
        case "chat.bsky.convo.defs#deletedMessageView":
            let deletedView = try singleValueContainer.decode(DeletedMessageView.self)
            self = .deletedMessageView(deletedView)
        default:
            throw DecodingError.dataCorruptedError(
                in: singleValueContainer,
                debugDescription: "Unable to decode MessageUnion with type: \(type)"
            )
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .messageView(let view):
            try container.encode(view)
        case .deletedMessageView(let view):
            try container.encode(view)
        }
    }

    var id: String {
        switch self {
        case .messageView(let view):
            return view.id
        case .deletedMessageView(let view):
            return view.id
        }
    }

    var sentAt: String {
        switch self {
        case .messageView(let view):
            return view.sentAt
        case .deletedMessageView(let view):
            return view.sentAt
        }
    }
}

struct MessageView: Codable, Identifiable {
    let id: String
    let rev: String
    let text: String?
    let facets: [Facet]?
    let embed: MessageEmbed?
    let sender: MessageSender
    let sentAt: String

    private enum CodingKeys: String, CodingKey {
        case id, rev, text, facets, embed, sender, sentAt
    }

    var sentAtDate: Date {
        ISO8601DateFormatter().date(from: sentAt) ?? Date()
    }
}

struct DeletedMessageView: Codable, Identifiable {
    let id: String
    let rev: String
    let sender: MessageSender
    let sentAt: String

    private enum CodingKeys: String, CodingKey {
        case id, rev, sender, sentAt
    }

    var sentAtDate: Date {
        ISO8601DateFormatter().date(from: sentAt) ?? Date()
    }
}

struct MessageEmbed: Codable {
    let record: EmbedRecord?

    struct EmbedRecord: Codable {
        let uri: String
        let cid: String
    }
}

// MARK: - Message Input

struct MessageInput: Codable {
    let text: String
    let facets: [Facet]?
    let embed: MessageEmbed?

    init(text: String, facets: [Facet]? = nil, embed: MessageEmbed? = nil) {
        self.text = text
        self.facets = facets
        self.embed = embed
    }
}

// MARK: - API Responses

struct ListConvosResponse: Codable {
    let cursor: String?
    let convos: [ConvoView]
}

struct GetConvoResponse: Codable {
    let convo: ConvoView
}

struct GetConvoForMembersResponse: Codable {
    let convo: ConvoView
}

struct SendMessageResponse: Codable {
    let id: String
    let rev: String
    let text: String?
    let facets: [Facet]?
    let embed: MessageEmbed?
    let sender: MessageSender
    let sentAt: String
}

struct GetMessagesResponse: Codable {
    let cursor: String?
    let messages: [MessageUnion]
}

// MARK: - Availability

struct GetConvoAvailabilityResponse: Codable {
    let status: String
}

File: app/src/main/java/com/cameronbanga/skyscraper/models/ChatModels.kt
package com.cameronbanga.skyscraper.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import java.time.Instant

// MARK: - Conversation Models

@JsonClass(generateAdapter = true)
data class ConvoView(
    val id: String,
    val rev: String,
    val members: List<ConvoMember>,
    val lastMessage: MessageUnion? = null,
    val muted: Boolean,
    val unreadCount: Int
) {
    fun otherMembers(currentUserDid: String?): List<ConvoMember> {
        return members.filter { !it.did.contains(currentUserDid ?: "") }
    }

    fun displayName(currentUserDid: String?): String {
        val others = otherMembers(currentUserDid)
        return when {
            others.size == 1 -> others.first().displayName ?: others.first().handle
            others.size > 1 -> others.take(2).joinToString(", ") { it.displayName ?: it.handle }
            else -> "Conversation"
        }
    }

    fun displayHandle(currentUserDid: String?): String? {
        val others = otherMembers(currentUserDid)
        return if (others.size == 1) {
            "@${others.first().handle}"
        } else null
    }

    fun avatarURL(currentUserDid: String?): String? {
        return otherMembers(currentUserDid).firstOrNull()?.avatar
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ConvoView) return false
        return id == other.id && rev == other.rev && unreadCount == other.unreadCount
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + rev.hashCode()
        result = 31 * result + unreadCount
        return result
    }
}

@JsonClass(generateAdapter = true)
data class ConvoMember(
    val did: String,
    val handle: String,
    val displayName: String? = null,
    val avatar: String? = null,
    val associated: Associated? = null
) {
    @JsonClass(generateAdapter = true)
    data class Associated(
        val chat: ChatAssociated? = null
    ) {
        @JsonClass(generateAdapter = true)
        data class ChatAssociated(
            val allowIncoming: String
        )
    }
}

// MARK: - Message Models

@JsonClass(generateAdapter = true)
data class MessageSender(
    val did: String
)

/**
 * Polymorphic message union - can be either a regular message or deleted message
 */
@JsonClass(generateAdapter = true)
data class MessageUnion(
    @field:Json(name = "\$type") val type: String,
    // MessageView fields
    val id: String? = null,
    val rev: String? = null,
    val text: String? = null,
    val facets: List<Facet>? = null,
    val embed: MessageEmbed? = null,
    val sender: MessageSender? = null,
    val sentAt: String? = null
) {
    val messageId: String get() = id ?: ""
    val messageSentAt: String get() = sentAt ?: ""

    fun toMessageView(): MessageView? {
        if (type == "chat.bsky.convo.defs#messageView") {
            return MessageView(
                id = id ?: return null,
                rev = rev ?: return null,
                text = text,
                facets = facets,
                embed = embed,
                sender = sender ?: return null,
                sentAt = sentAt ?: return null
            )
        }
        return null
    }

    fun toDeletedMessageView(): DeletedMessageView? {
        if (type == "chat.bsky.convo.defs#deletedMessageView") {
            return DeletedMessageView(
                id = id ?: return null,
                rev = rev ?: return null,
                sender = sender ?: return null,
                sentAt = sentAt ?: return null
            )
        }
        return null
    }
}

@JsonClass(generateAdapter = true)
data class MessageView(
    val id: String,
    val rev: String,
    val text: String? = null,
    val facets: List<Facet>? = null,
    val embed: MessageEmbed? = null,
    val sender: MessageSender,
    val sentAt: String
) {
    val sentAtDate: Instant
        get() = try {
            Instant.parse(sentAt)
        } catch (e: Exception) {
            Instant.now()
        }
}

@JsonClass(generateAdapter = true)
data class DeletedMessageView(
    val id: String,
    val rev: String,
    val sender: MessageSender,
    val sentAt: String
) {
    val sentAtDate: Instant
        get() = try {
            Instant.parse(sentAt)
        } catch (e: Exception) {
            Instant.now()
        }
}

@JsonClass(generateAdapter = true)
data class MessageEmbed(
    val record: EmbedRecord? = null
) {
    @JsonClass(generateAdapter = true)
    data class EmbedRecord(
        val uri: String,
        val cid: String
    )
}

// MARK: - Message Input

@JsonClass(generateAdapter = true)
data class MessageInput(
    val text: String,
    val facets: List<Facet>? = null,
    val embed: MessageEmbed? = null
)

// MARK: - API Response Models

@JsonClass(generateAdapter = true)
data class ListConvosResponse(
    val cursor: String? = null,
    val convos: List<ConvoView>
)

@JsonClass(generateAdapter = true)
data class GetConvoResponse(
    val convo: ConvoView
)

@JsonClass(generateAdapter = true)
data class GetConvoForMembersResponse(
    val convo: ConvoView
)

@JsonClass(generateAdapter = true)
data class SendMessageResponse(
    val id: String,
    val rev: String,
    val text: String? = null,
    val facets: List<Facet>? = null,
    val embed: MessageEmbed? = null,
    val sender: MessageSender,
    val sentAt: String
)

@JsonClass(generateAdapter = true)
data class GetMessagesResponse(
    val cursor: String? = null,
    val messages: List<MessageUnion>
)

@JsonClass(generateAdapter = true)
data class GetConvoAvailabilityResponse(
    val status: String
)
